### PR TITLE
Add ConcatenatedDataset

### PR DESCRIPTION
**Pull request by Hakuyume**
I added a new dataset wrapper, `ConcatenatedDataset`. This is similar to `SubsetDataset` but this dataset represents a union set of some base datasets.

### CODE REVIEW

File: chainer/datasets/concatenated_dataset.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class ConcatenatedDataset(object):
```
Comment (niboshi) [inline chainer/datasets/concatenated_dataset.py]: How about implementing it with DatasetMixin? Handling indices will be much easier and nd.array index will be automatically supported.

File: chainer/datasets/concatenated_dataset.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    Args:
+        datasets: Underlying datasets.
+
+    """
+
+    def __init__(self, *datasets):
+        self._datasets = datasets
+
+    def __len__(self):
+        return sum(map(len, self._datasets))
```
Comment (niboshi) [inline chainer/datasets/concatenated_dataset.py]: Rewrite with generator, i.e. `sum(len(_) for _ in self._datasets)`

File: chainer/datasets/concatenated_dataset.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    This dataset wraps some base datasets and works as a concatenated dataset.
+    For example, if a base dataset with 10 samples and
+    another base dataset with 20 samples are given, this dataset works as
+    a dataset which has 30 samples.
+
+    .. note::
+
+        The index used to access data is either an integer or a slice. If it
+        is a slice, the base dataset is assumed to return a list of outputs
+        each corresponding to the output of the integer indexing.
```
Comment (niboshi) [inline chainer/datasets/concatenated_dataset.py]: Is this note needed?

File: docs/source/reference/datasets.rst [hunk 4]
```diff
@@ -21,11 +21,13 @@ These concrete examples use the downloading utilities in the :mod:`chainer.datas
...
 ----------------
 
-General datasets are further divided into three types.
+General datasets are further divided into four types.
 
 The first one is :class:`DictDataset` and :class:`TupleDataset`, both of which combine other datasets and introduce some structures on them.
 
-The second one is :class:`SubDataset`, which represents a subset of an existing dataset. It can be used to separate a dataset for hold-out validation or cross validation. Convenient functions to make random splits are also provided.
+The second one is :class:`ConcatenatedDataset` and :class:`SubDataset`.
+:class:`ConcatenatedDataset` represents a union of existing datasets. It can be used to merge datasets and make a large dataset.
```
Comment (niboshi) [inline docs/source/reference/datasets.rst]: * "union" is misleading. How about simply "concatenation"?
* large -> larger

File: tests/chainer_tests/datasets_tests/test_concatenated_dataset.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import unittest
+
+import numpy as np
+
+from chainer.datasets import ConcatenatedDataset
+from chainer import testing
+
+
+class TestConcatenatedDataset(unittest.TestCase):
```
Comment (niboshi) [inline tests/chainer_tests/datasets_tests/test_concatenated_dataset.py]: Please add these tests:
* empty list of base dataset is given (`ConcatenatedDataset()`)
* only a single base dataset is given (`ConcatenatedDataset(dataset0)`)
* Some/all base datasets have 0 samples

**Comment by yuyu2172** [comment]
We should have a description of the argument in the docstring using `Args:`.

**Comment by Hakuyume** [comment]
@yuyu2172 Thank you for pointing out.

**Comment by niboshi** [comment]
@Hakuyume  Thank you for fix, and @yuyu2172 thank you for checking docstring.
LGTM, I will merge this PR after confirming test success.

**Comment by niboshi** [comment]
Confirmed tests to pass.
Thank you very much!

**Pull request merged**